src/csharp/Grpc.Core/CallOptions.cs

```diff
@@ -36,6 +37,7 @@ public struct CallOptions
         ContextPropagationToken propagationToken;
         CallCredentials credentials;
         CallFlags flags;
+        Dictionary<object, object> items;
```

Comment:
I'm a huge fan of object versus string, sometime I like to store things via by type.

Get<T>() for example.

Reply:
I'm concerned that this already-large struct is becoming huge. The threading is also a concern... `Dictionary<,>` isn't thread-safe - what are the expected uses here, and are we satisfied they won't cause problems?

Reply:
@jskeet The use-case is to share some call-specific data between the interceptors and potentially the call invoker. (The closest .NET analogy I have in mind is `HttpContext.Context`.) I initially thought Metadata headers should be abused for this purpose, but it can be too inconvenient. For example, one use case is associating a `User` object (or any other state information) via an interceptor that can then be pulled and injected into the RPC.  Another design I contemplated with is replacing the dictionary with a single `object Tag`, however, from a struct size point of view, it has the same effect. It will alleviate the threading concerns though, perhaps at a cost to usability.

Reply:
I understand the use case for "items", but I share @jskeet 's concerns.  

Idea - could we postpone adding the "items" field and later add it without breaking the API? Adding the basic functionality for interceptors in this PR and having one or two "improvements" PR might actually be productive approach.

Reply:
Happy to move `items` to a different PR, but this is an important requirement it seems, and we need to tackle it soon anyway. Metadata strings are a really ugly approach.

Another idea is to introduce an `InterceptorContext` which holds interceptor state and pass it along in the continuation chain. However, the downside is there is no clear way to share that information with the final `CallInvoker` (which might or might not be necessary). If we are going to do this, we can't do it in another PR, because that is a core API change in the interceptor itself (adding another argument to interception callbacks. and the continuation delegate. I'd really like to know what @jskeet thinks about this alternative solution.